### PR TITLE
Hide attached DWH database details

### PR DESCRIPTION
**Pull request by devurandom**
== Goal ==

Hide attached DWH database details from anyone incl. admins:
* Do not show them in the UI
* Do not permit to change them
* Do not serialize them

The aim is that customers cannot gain access to (parts of) credentials,
and they cannot break a feature they are paying for by changing
connection details.

Database details are already generally excluded from H2 dump snapshots
(see `metabase.cmd.copy/*copy-h2-database-details*`), thus nothing
changes there.

== Implementation ==

The Metabase backend already contains provisions in the implementation
of `metabase.models.interface/to-json` for `:model/Database` to hide the
`details` of the database in HTTP responses, if the user lacks write
permission on the database.  We utilize this by adding an
`is_attached_dwh` column to the `database` table and rejecting
`metabase.models.interface/can-write?` when this flag is enabled.  In
the "admin" UI, we show a replacement text instead of the edit form when
the flag is set.  (It might be correct to show this whenever `details`
is absent.  See below for possible follow-up work.)

However, several sections of the frontend code expected the `details`
field to always be present.  In order to make `details` optional, as the
backend seems to handle it, we fix the respective code to treat this
case in the way that appears appropriate in the context.

== How to test ==

=== New behaviour ===

Setting the `is_attached_dwh` field hides the database details:

1. Configure a database as described in https://www.metabase.com/docs/latest/configuring-metabase/config-file#databases.
   - In addition to the fields you would normally set, also set
     `is_attached_dwh: true`.
   - This also works when adding this flag to a database that previously
     did not have this flag set.
2. Start your Metabase instance.
3. Verify the database shows up in the "admin" section
   (`/admin/databases`).
4. Verify that clicking the database to see its details only reveals
   "This database cannot be modified."
5. Verify that responses from the backend do not include a `details`
   field for this database.

=== Original behaviour ===

Behaviour without setting the `is_attached_dwh` field is unchanged:

1. Configure a database as described in https://www.metabase.com/docs/latest/configuring-metabase/config-file#databases.
   - Only set the fields you would normally set.  Do not set
     `is_attached_dwh` (or set it to `false`).
2. Start your Metabase instance.
3. Verify the database shows up in the "admin" section
   (`/admin/databases`).
4. Verify that clicking the database to see its details only reveal the
   regular edit form, showing connection fields like `host`, `user`,
   etc. with the values you configured.

== How this will be rolled out ==

1. Upgrade existing Metabase Cloud instances with data warehouse to a
   Metabase version that supports `is_attached_dwh`.
2. Set `is_attached_dwh` in the database section of the config file for
   Metabase Cloud instances with a data warehouse.

== Possible follow-up work ==

In https://github.com/metabase/metabase/issues/25715, absent
`database.details` was identified as a bug.  Since then, `details` was
made `NOT NULL` in the application database, so this bug can no longer
occur.  However, today backend responses can be missing the `details`
field, if the current user lacks write permission to the database
setting (see above).  Fully re-evaluating #25715 in this context is
outside the scope of this PR.

Closes: https://github.com/metabase/harbormaster/issues/5051

### CODE REVIEW

File: src/metabase/models/database.clj [hunk 1]
```diff
@@ -434,14 +444,18 @@
...
-               :description :engine :is_audit :is_full_sync :is_on_demand :is_sample :metadata_sync_schedule :name
+               :description :engine :is_audit :is_attached_dwh :is_full_sync :is_on_demand :is_sample :metadata_sync_schedule :name
                :points_of_interest :refingerprint :settings :timezone :uploads_enabled :uploads_schema_name
                :uploads_table_prefix]
    :skip      [;; deprecated field
                :cache_ttl]
    :transform {:created_at          (serdes/date)
                ;; details should be imported if available regardless of options
-               :details             {:export #(if include-database-secrets % ::serdes/skip) :import identity}
+               :details             {:export (fn [{:keys [is_attached_dwh] :as db}]
```
Comment (crisptrutski) [inline src/metabase/models/database.clj:454]: The argument coming to this function will be the value of `:details`, not the entire record. You would need to use the dynamic variable `serdes/*currrent* to read `is_attached_dwh` from.

File: src/metabase/models/database.clj [hunk 2]
```diff
@@ -434,14 +444,18 @@
...
+                                               (if (and include-database-secrets (not is_attached_dwh))
+                                                 db
```
Comment (crisptrutski) [inline src/metabase/models/database.clj:456]: This return value should be just the details

File: src/metabase/models/database.clj [hunk 3]
```diff
@@ -311,6 +317,10 @@
 
 (t2/define-after-update :model/Database
   [database]
+  ;; Like for `is_audit`, we do not support removing this flag:
+  (when (:is_attached_dwh database)
+    (let [admin-group (perms-group/admin)]
+      (data-perms/set-database-permission! admin-group database :perms/manage-database :no)))
```
Comment (crisptrutski) [inline src/metabase/models/database.clj:323]: Some questions / things to think about.

1. When do we anticipate this property being changed? Is this related to 2?
2. What's our plan for enabling this property / stripping these permissions for existing instances? 
4. Do we not want a hook to prevent `:is_attached_dwh` being disabled?
5. Even allowing customers to enable it via the API seems like something we wouldn't want.

**Review by crisptrutski — COMMENTED** [review]
Left some comments on the lifecycle and serialization. The permission side of things seems sensible, but I leave that for the relevant experts to review.

File: src/metabase/models/database.clj (hunk 2 above)
Comment (devurandom) [reply src/metabase/models/database.clj]: https://github.com/metabase/metabase/pull/47247/commits/072fb725ee1392eb5b59382290eebc39ce1df60c

File: src/metabase/models/database.clj (hunk 1 above)
Comment (devurandom) [reply src/metabase/models/database.clj]: Thanks! https://github.com/metabase/metabase/pull/47247/commits/072fb725ee1392eb5b59382290eebc39ce1df60c

File: src/metabase/models/database.clj (hunk 3 above)
Comment (crisptrutski) [reply src/metabase/models/database.clj]: Thanks, sounds solid

**Review by crisptrutski — APPROVED** [review]
Serialization portion LGTM

File: src/metabase/models/database.clj [hunk 4]
```diff
@@ -88,11 +88,14 @@
   (mi/superuser?))
 
 (defmethod mi/can-write? :model/Database
+  ;; Lack of permission to change database details will also exclude the `details` field from the HTTP response,
+  ;; cf. the implementation of [[metabase.models.interface/to-json]] for `:model/Database`.
   ([instance]
    (mi/can-write? :model/Database (u/the-id instance)))
```
Comment (noahmoss) [inline src/metabase/models/database.clj:94]: Could you change this so that the 1-arity version checks `:is_attached_dwh` on the `instance` directly if it exists, rather than recursively calling into the 2-arity version and then making an unnecessary DB call?

It seems like the 1-arity version of `can-write?` is what's used in most places in `metabase.api.database` so this would be a helpful optimization, especially when we're fetching a large list of DBs.

**Review by noahmoss — APPROVED** [review]
Looks good aside from one small refactor request

File: src/metabase/models/database.clj (hunk 4 above)
Comment (devurandom) [reply src/metabase/models/database.clj]: Is https://github.com/metabase/metabase/pull/47247/commits/98ca3b19ed3439d471950252800e28d153a64eae (close to) what you had in mind?

**Review by npfitz — APPROVED** [review]
One thing I noticed is that in EE, if you have a database that is a dwh, it will not show up in the list of DBs in the admin page. This is because in EE only, we pass the query param `exclude_uneditable_details: true`, which excludes it from the list. I'm not sure if that is intentional or not

Aside from that, the code more less looks okay. Any boolean checks will just return false if the details aren't present, which should be a valid option, and the change in `schema.ts` shouldn't matter because it only gets called when you submit a database form, which you can't do for a DWH anyways.

File: src/metabase/models/database.clj (hunk 4 above)
Comment (noahmoss) [reply src/metabase/models/database.clj]: Yeah that looks fine. Is `:is_attached_dwh` is added to DBs automatically when they're fetched in the `metabase.api.database` endpoints?

**Review by iethree — CHANGES_REQUESTED** [review]
Overall looks good, a couple changes needed though

File: src/metabase/models/database.clj (hunk 4 above)
Comment (devurandom) [reply src/metabase/models/database.clj]: Yes, that works automatically. `is_attached_dwh` appears in the HTTP response for `GET /api/database/2`.

**Comment by devurandom** [comment]
> One thing I noticed is that in EE, if you have a database that is a dwh, it will not show up in the list of DBs in the admin page. This is because in EE only, we pass the query param `exclude_uneditable_details: true`, which excludes it from the list. I'm not sure if that is intentional or not

Thanks!  Even in EE admins should be able to see that their DWH exists, they just should not be able to edit it.  I fixed that in https://github.com/metabase/metabase/pull/47247/commits/bf2dbdd72b731b5ec82ae376591f5becba160d13.

I did not have to change `GET /:id`, because we do not pass `exclude_uneditable_details` when requesting that.  (`PLUGIN_FEATURE_LEVEL_PERMISSIONS.databaseDetailsQueryProps` is only used in `frontend/src/metabase/admin/databases/containers/DatabaseListApp.jsx` and `frontend/src/metabase/admin/databases/components/DatabaseList/DatabaseList.jsx`.)

File: frontend/src/metabase/admin/databases/containers/DatabaseEditApp.tsx [hunk 5]
```diff
@@ -170,13 +170,17 @@ function DatabaseEditApp(props: DatabaseEditAppProps) {
...
                 <DatabaseEditContent>
                   <DatabaseEditForm>
-                    <DatabaseForm
-                      initialValues={database}
-                      isAdvanced
-                      onSubmit={handleSubmit}
-                      setIsDirty={setIsDirty}
-                      autofocusFieldName={autofocusFieldName}
-                    />
+                    {editingExistingDatabase && database.is_attached_dwh ? (
```
[the remark this answers is not in the document]
Comment (devurandom) [reply frontend/src/metabase/admin/databases/containers/DatabaseEditApp.tsx]: Thanks! I fixed the sidebar in https://github.com/metabase/metabase/pull/47247/commits/7715b5b3bb828731c936f6eaa74ae36fdbc1fc68.

File: frontend/src/metabase/admin/databases/containers/DatabaseEditApp.tsx [hunk 6]
```diff
@@ -170,13 +170,17 @@ function DatabaseEditApp(props: DatabaseEditAppProps) {
...
+                      <div>This database cannot be modified.</div>
```
[the remark this answers is not in the document]
Comment (devurandom) [reply frontend/src/metabase/admin/databases/containers/DatabaseEditApp.tsx]: Thanks, fixed in https://github.com/metabase/metabase/pull/47247/commits/9fec62c1bb1c8f994955f130017045021d3ae2fd.

File: e2e/test/scenarios/admin/databases.cy.spec.js [hunk 7]
```diff
@@ -489,6 +489,23 @@ describe("scenarios > admin > databases > exceptions", () => {
...
+        res.body.is_attached_dwh = true;
+      });
+    }).as("loadDatabase");
+
+    cy.visit("/admin/databases/1");
+    cy.wait("@loadDatabase");
+
+    cy.get("nav").should("contain", "Metabase Admin");
+    cy.get("span").contains(/Sample Database/i);
+    cy.get("div").contains("This database cannot be modified.");
```
[the remark this answers is not in the document]
Comment (devurandom) [reply e2e/test/scenarios/admin/databases.cy.spec.js]: Thanks for the guidance!  Fixed in https://github.com/metabase/metabase/pull/47247/commits/4a46aa3de515d642351069b9146be3321e584bac.

**Review by iethree — APPROVED** [review]
✨

**Comment by npfitz** [comment]
Change to the endpoint looks good to me! Thanks @devurandom
